sound/soc/soc-core.c

```diff
@@ -1656,9 +1656,11 @@ static void soc_check_tplg_fes(struct snd_soc_card *card)
 			dai_link->platforms->name = component->name;
 
 			/* convert non BE into BE */
-			dai_link->no_pcm = 1;
-			dai_link->dpcm_playback = 1;
-			dai_link->dpcm_capture = 1;
+			if (!dai_link->no_pcm) {
+				dai_link->no_pcm = 1;
+				dai_link->dpcm_playback = 1;
+				dai_link->dpcm_capture = 1;
+			}
```

Comment:
@bardliao  why we get BE here ?   this function only deal with FE in topology.

Reply:
@RanderWang If soc_check_tplg_fes find a matched machine (like sof_sdw), it will overwrite all preset links. See [for_each_card_prelinks(card, i, dai_link)](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-core.c#L1640)

Reply:
so soc_check_tplg_fes doesn't only check fes.  why force dpcm_playback & dpcm_capture to 1 ? I don't get it.

Reply:
@RanderWang  @bardliao i think I'm at fault for this

```
commit 218fe9b7ec7f32c10a07539365488d80af7b0084
Author: Daniel Baluta <daniel.baluta@nxp.com>
Date:   Wed Dec 4 17:13:33 2019 +0200

    ASoC: soc-core: Set dpcm_playback / dpcm_capture
    
    When converting a normal link to a DPCM link we need
    to set dpcm_playback / dpcm_capture otherwise playback/capture
    streams will not be created resulting in errors like this:
    
    [   36.039111]  sai1-wm8960-hifi: ASoC: no backend playback stream
    
    Fixes: a655de808cbde ("ASoC: core: Allow topology to override machine driver FE DAI link config")
```

I think the proper fix would be to check if the original link has playback OR capture capability and update only that capability instead of updating both playback and capture.